pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HdfsSegmentFetcher.java

```diff
@@ -0,0 +1,103 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Created by jamesshao on 9/14/17.
+ */
+public class HdfsSegmentFetcher implements SegmentFetcher {
+
+  private static final String PRINCIPLE = "hadoop.kerberos.principle";
+  private static final String KEYTAB = "hadoop.kerberos.keytab";
+  private static final String HADOOP_CONF_PATH = "hadoop.conf.path";
+  private static final Logger LOGGER = LoggerFactory.getLogger(HdfsSegmentFetcher.class);
+  private FileSystem fs = null;
+
+  @Override
+  public void init(Map<String, String> configs) {
+    String hadoopConfPath = configs.get(HADOOP_CONF_PATH);
+    Configuration hadoopConf = getConf(hadoopConfPath);
+    authenticate(hadoopConf, configs);
+    try {
+      fs = FileSystem.get(hadoopConf);
+      LOGGER.debug("successfully initialized hdfs segment fetcher");
+    } catch (IOException e) {
+      LOGGER.error("failed to initialized the hdfs", e);
+    }
+  }
+
+  private Configuration getConf(String hadoopConfPath) {
+    Configuration hadoopConf = new Configuration();
+    if (Strings.isNullOrEmpty(hadoopConfPath)) {
+      LOGGER.warn("no hadoop conf path is provided, will rely on default config");
+    } else {
+      hadoopConf.addResource(new Path(hadoopConfPath, "core-site.xml"));
+      hadoopConf.addResource(new Path(hadoopConfPath, "hdfs-site.xml"));
+    }
+    return hadoopConf;
+  }
+
+  private void authenticate(Configuration hadoopConf, Map<String, String> configs) {
+    String principal = configs.get(PRINCIPLE);
+    String keytab = configs.get(KEYTAB);
+    if (!Strings.isNullOrEmpty(principal) && !Strings.isNullOrEmpty(keytab)) {
+      UserGroupInformation.setConfiguration(hadoopConf);
+      if (UserGroupInformation.isSecurityEnabled()) {
+        try {
+          if (!UserGroupInformation.getCurrentUser().hasKerberosCredentials()
+              || !UserGroupInformation.getCurrentUser().getUserName().equals(principal)) {
+            LOGGER.info("Trying to authenticate user [%s] with keytab [%s]..", principal, keytab);
+            UserGroupInformation.loginUserFromKeytab(principal, keytab);
+          }
+        }
+        catch (IOException e) {
+          throw new RuntimeException(String.format("Failed to authenticate user principal [%s] with keytab [%s]",
+                  principal, keytab), e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void fetchSegmentToLocal(String uri, File tempFile) throws Exception {
+    LOGGER.debug("starting to fetch segment from hdfs");
+    try {
+      if (fs == null) {
+        LOGGER.error("uninitialized fs for fetching data from hdfs");
+        throw new RuntimeException("failed to get hdfs client");
+      }
+      
+      fs.copyToLocalFile(new Path(uri), new Path(tempFile.toURI()));
+      LOGGER.debug("copied {} from hdfs to {} in local for size {}", uri, tempFile.getAbsolutePath(), tempFile.length());
+    } catch(Exception ex) {
```

Comment:
- check for existing of the file first
- after downloading, check for size/crc to ensure the data is correctly copied
Are there different types of exceptions? If there is a connection break, It will be good to retry.

Reply:
I researched a bit on crc calculation for hadoop/local fs, it seems that according to this stackoverflow answer https://stackoverflow.com/a/14564532 it is tricky to implement the crc validation across different file systems. I think for the sake of simplicity we can skip this part as we are not doing that for other fetchers as well. 

For checking local file existence, I think this might be dangerous as they might be partial copy results. to verify that file content is correct we will need to do checksum verification or other things. This leads us back to my previous comments